src/parser/jobs/smn/modules/MissingDoTs.js

```diff
@@ -0,0 +1,135 @@
+import {t} from '@lingui/macro'
+import {Trans, Plural} from '@lingui/react'
+import React from 'react'
+
+import {ActionLink, StatusLink} from 'components/ui/DbLink'
+import {RotationTable} from 'components/ui/RotationTable'
+import ACTIONS from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import Module from 'parser/core/Module'
+import {TieredSuggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+
+// In potency
+const MISSING_DOT_SEVERITY = {
+	1: SEVERITY.MEDIUM,
+	1000: SEVERITY.MAJOR,
+}
+
+// Statuses that need to be up for Fester, Bane, and Ruins to do the most
+const SMN_DOT_STATUSES = [
+	STATUSES.BIO_III.id,
+	STATUSES.MIASMA_III.id,
+]
+
+export default class MissingDoTs extends Module {
+	static handle = 'missingdots'
+	static title = t('smn.missingdots.title')`Missing DoTs`
+	static dependencies = [
+		'enemies',
+		'suggestions',
+		'timeline',
+	]
+
+	_badDotReqCasts = {
+		[ACTIONS.FESTER.id]: {
+			potPerDot: 100,
+			badCastCounts: [0, 0],
+		},
+		[ACTIONS.RUIN_II.id]: {
+			potPerDot: 40,
+			badCastCounts: [0, 0],
+		},
+		[ACTIONS.RUIN_III.id]: {
+			potPerDot: 50,
+			badCastCounts: [0, 0],
+		},
+		[ACTIONS.RUIN_IV.id]: {
+			potPerDot: 70,
+			badCastCounts: [0, 0],
+		},
+	}
+	_missingDotWindows = []
+	_currentMissingDotWindow = {
+		timestamp: 0, //placeholder
+		casts: [],
+	}
+
+	constructor(...args) {
+		super(...args)
+
+		this.addHook('complete', this._onComplete)
+		this.addHook('cast', {
+			by: 'player',
+			abilityId: [ACTIONS.FESTER.id, ACTIONS.RUIN_II.id, ACTIONS.RUIN_III.id, ACTIONS.RUIN_IV.id],
+		}, this._onDotReqCast)
+	}
+
+	_onComplete() {
+		// Suggestion for fester/ruin casts without both dots
+		// TODO: add exception for initial cast (and possibly first cast after return) where Ruins are being cast before Tri-D
```

Comment:
mm, that could be an interesting one to try and account for

Reply:
the pre-cast could probably be filtered out by timestamp < 2s or some suitably small number.  Guides seem to be suggesting clipping to cast Tri-D before the next Ruin 2, so you wouldn't need to catch that one.

Following that reasoning, you would probably cast Tri-D on return before the first post-return GCD.  If Tri-D isn't available, you should be doing B3/M3 hardcasts first, so that case wouldn't be an issue.

There may also be some issues around trance resets (boss returns, dots expired, trance ready, Tri-D still on cooldown), but i don't know what the expected action will be there, so i didn't try to code for it.